rhel/vex/parser.go

```diff
@@ -183,31 +183,46 @@ type creator struct {
 	rc                 *repoCache
 }
 
-// WalkRelationships attempts to resolve a relationship until we have a package product_id and
-// a repo product_id. Relationships can be nested. If the pkgID and the repoID are the same we
-// either found no relationship or a relationship where both ends are pointing to the same
-// product_id, either way we don't have enough data to create a vulnerability.
-func walkRelationships(productID string, doc *csaf.CSAF) (string, string, error) {
-	pkgID, repoID := extractProductNames(productID, productID, doc)
-	if pkgID == repoID {
-		return "", "", fmt.Errorf("could not extract a distict pkgID and repoID from %q", productID)
+// WalkRelationships attempts to resolve a relationship until we have a package product_id,
+// a repo product_id and possibly a package module product_id. Relationships can be nested.
+// If we don't get an initial relationship or we don't get two component parts we cannot
+// create a vulnerability. We never see more than 3 components in the wild but if we did
+// we'd assume the component next to the repo product_id is the package module.
+func walkRelationships(productID string, doc *csaf.CSAF) (string, string, string, error) {
```

Comment:
hoping you can help me walk through this example:

I'm looking at https://security.access.redhat.com/data/csaf/v2/vex/2023/cve-2023-47038.json and I'm interested in `perl-Carp`.

I can find "AppStream-8.10.0.GA:perl:5.32:8100020240314121426:9fe1d287:perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch" in the "fixed" section, so I use that as `productID` for this function. Then, I look for the "default_component_of" with this product ID. This gives me a `prodRel` looking like

```
ProductRef:                 "perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch",
RelatesToProductRef: "AppStream-8.10.0.GA:perl:5.32:8100020240314121426:9fe1d287",
```

Then inside `extractProductNames` I look for "default_component_of" with `product_id` of `perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch`.

This is where I get confused. I don't see this existing anywhere. Can you tell me if I'm reading this wrong?

Reply:
`perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch` exists as a product in the product branches, that's the end of the road for that product, no more relationships to resolve. `AppStream-8.10.0.GA:perl:5.32:8100020240314121426:9fe1d287` on the other hand refers to another relationship: 
```json
        "product_reference": "perl:5.32:8100020240314121426:9fe1d287",
        "relates_to_product_reference": "AppStream-8.10.0.GA"
```
Both of these refer to a product in the product branches. Does this make sense?

Reply:
so `comps = extractProductNames(prodRel.ProductRef, prodRel.RelatesToProductRef, comps, doc)` will return ```
comps = []string{"perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch", "perl:5.32:8100020240314121426:9fe1d287", "AppStream-8.10.0.GA"}
```
?

Reply:
package: `perl-Carp-0:1.50-439.module+el8.10.0+21354+3ad137bb.noarch`
module: `perl:5.32:8100020240314121426:9fe1d287`
repo: `AppStream-8.10.0.GA`

Reply:
Yeah, added a unit-test to confirm